### PR TITLE
Skills : Validate user skills by subcategory in gradebook - refs BT#19823

### PR DESCRIPTION
**Pull request by christianbeeznest**

### CODE REVIEW

File: main/inc/lib/skill.lib.php [hunk 1]
```diff
@@ -1208,6 +1235,41 @@ public function addSkillToUser(
...
+            );
+            $alleval = $category->get_evaluations($userId, false, $category->get_course_code(),
+                $category->get_session_id());
+            $alllink = $category->get_links($userId, true, $category->get_course_code(), $category->get_session_id());
+            $datagen = new GradebookDataGenerator($subCategories, $alleval, $alllink);
+            $gradeResult = $datagen->get_data();
+            foreach ($gradeResult as $data) {
+                /** @var AbstractLink $item */
+                $item = $data[0];
+                if ('Category' === get_class($item)) {
```
Comment (AngelFQC) [inline main/inc/lib/skill.lib.php:1263]: `Category::class === get_class($item)` or `$item instanceof Category::class`

File: main/install/configuration.dist.php [hunk 2]
```diff
@@ -2156,6 +2156,11 @@
 // User extra fields to be check on user edition to generate a specific process if it was modified
 //$_configuration['user_edition_extra_field_to_check'] = 'ExtrafieldLabel';
 
+// Enable skills in subcategory to work independant on assignement
+// Require DB changes:
+// ALTER TABLE gradebook_category ADD allow_skills_by_subcategory tinyint(1) NOT NULL DEFAULT '1';
```
Comment (AngelFQC) [inline main/install/configuration.dist.php:2161]: Add the property to the GradebookCategory entity but disabled.

**Review by AngelFQC — CHANGES_REQUESTED** [review]
Mejor hacer uso de entidades que queries SQL

File: main/inc/lib/skill.lib.php (hunk 1 above)
Comment (christianbeeznest) [reply main/inc/lib/skill.lib.php]: corregido en el ultimo commit

File: main/install/configuration.dist.php (hunk 2 above)
Comment (christianbeeznest) [reply main/install/configuration.dist.php:2161]: agregado en el ultimo commit

**Pull request merged**